octoprint_enclosure/__init__.py

```diff
@@ -25,13 +25,16 @@
 from .getPiTemp import PiTemp
 import struct
 
+## Import modifications for Hardware PWM using PiGPIO
+import pigpio

```

Comment:
this import forces all users of this plugin to install/import pigpio, even though the majority of the users won't use it.

Reply:
Right.  This also assumes that that the only case for using hardware pwm is a fan linked to a temperature sensor.  A better way to do this is change the UI (ninja2?) to have the user choose either hardware pwm or software pwm as an output type and not just pwm.  I'm willing to figure out how to do that if there is enough interest,

Reply:
I agree with @wmonzel on this one, lack of hardware PWM support is a problem for me unfortunately, and means I have to resort to a separate plugin just to run the fans, meaning I lose the benefits of controlling the fan based on temperature, which is critical for maintaining a controlled temperature for printing within the enclosure.

Personally I think adding `pigpio` as an optional dependency would be nice, so those that need hardware PWM can install it, while those who don't need it, don't have to install additional software on the host. If it's possible to add support for this without making `pigpio` a required dependency, that would be absolutely amazing.

Reply:
I am having the same issues with a Noctua fan and also agree with @wmonzel that having a hardware PWM option would be helpful as I cannot get the current PWM option to work reliability at the moment.

Reply:
I've tested with pigpio and it's quite heavyweight using a chunk of resources scanning all the GPIO pins many times per second. I'm going to try and see if a lighter weight alternative is suitable: https://github.com/Pioreactor/rpi_hardware_pwm and allow this to be configured via the UI so that only users that want to use a hardware PWM will need to install this extra module.  Not sure it will work, but it might address the concerns raised in the comment on this PR.

Reply:
@thijstriemstra @bilde2910 @wmonzel I hope you don't mind. I've implemented an alternative to this PR which I believe addresses the concerns previously raised. See #474

Reply:
@shortbloke @thijstriemstra @bilde2910 I really do appreciate you doing this.  I'm trying to test your code and have run into issues installing rpi-hardware-pwm.  The installation fails with the error:
**Could not find a version that satisfies the requirement rpi-hardware-pwm (from versions: )
No matching distribution found for rpi-hardware-pwm**
The installation command in Pioreactor uses pip3 (sudo pip3 install rpi-hardware-pwm) and fails with pip3: command not found.  
Trying to find a fix I've discovered that installing using sudo pip install to install packages is considered risky, and the package does not appear in the pypi.org or piwheels.org indexes.  
Any ideas?

Reply:
@wmonzel It sounds like your system is configured to have Python 2.7 as the default version of pip. This library needs Python 3. Assuming you have Python3 installed, check you have the Python3 version installed: `sudo apt-get install python3-pip`  and/or see if you have pip but need to update it: `python3 -m pip install --upgrade pip`

By installing a python package with `sudo` it makes it available to everything, if you have OctoPrint running inside a venv then you can activate that and just install the package without `sudo`.

The rpi-hardware-pwm package is on pypi.org: https://pypi.org/project/rpi-hardware-pwm/

Reply:
@shortbloke Thanks for your help.  Changing the default version of Python did allow me to install rpi-hardware-pwm.  But setting update-alternatives to auto mode (with python3.7 as the default) renders the MCP9808 temperature sensor unreadable.  It appears to be dependent on python2.x since manually setting to python2.x fixes the temperature sensor but breaks rpi-hardware-pwm. 

I'm learning as I go...

Reply:
Which library is that sensor using and I'll take a look to see if we can fix that too.
FYI Octoprint will require Python3 as of version 1.8 - https://octoprint.org/blog/2022/01/31/octoprint-1.8.0-will-require-python-3/

Reply:
@shortbloke There already is an open issue for this: #302  It appears that multiple sensors are affected.  After doing some digging, I've come to the conclusion that using pigpio is a better choice for the following reasons:
<ul>
 <li>It turns out that the CPU load issue can be fixed.  If scanning inputs is not required you can add the -m option (disable sampling) to ExecStart=/usr/bin/pigpiod -1 by editing /lib/systemd/system/pigpiod.service.   On my Pi4 CPU went from around 8% to 0% for pigpiod.  The sampling rate can also be reduced.  
 <li>Pigpio is currently used by the Easy Servo and HardwarePWM plugins.  Using HardwarePWM in conjunction with Enclosure gives you an easy way manually override the fan speed as needed.  I find this very useful. 
<li> Pigpio and HardwarePWM both appear to be mature and well supported.  rpi-hardware-pwm is fairly new and released specifically to support the Pioreactor code.   
<li> Pigpio could potentially be used to fully replace rpi.gpio
</ul>
This pull request needs a lot more work before it is ready for release.  I'm considering closing it and lending my support to your effort instead.  How do you feel about switching to pigpio?

Reply:
Sorry for the delay in responding @wmonzel I agree pigpio has its advantages. There is a disadvantage in that is requires a little bit more setup and modification to the installed service. I was aiming to keep it simple and not place any requirements on those that didn't want to use any hardware pwm. 
I could look to extend my PR further and offer support for multiple hardware pwm libraries, though that would create some added complexity in the UI. 
Given the lack of progress of either of our PRs, it seems demand for this support is rather low ;-)